art/attacks/poisoning/sleeper_agent_attack.py

```diff
@@ -0,0 +1,389 @@
+        :param clip_values: The range of the input features to the classifier.
+        :param verbose: Show progress bars.
+        :indices_target: The indices of training data having target label.
+        :patching_strategy: Patching strategy to be used for adding trigger, either random/fixed.
+        :selection_strategy: Selection strategy for getting the indices of
+                             poison examples - either random/maximum gradient norm.
+        :retraining_factor: The factor for which retraining needs to be applied.
+        :model_retrain: True, if retraining has to be applied, else False.
+        :model_retraining_epoch: The epochs for which retraining has to be applied.
+        :patch: The patch to be applied as trigger.
+        """
+        super().__init__(
+            classifier,
+            percent_poison,
+            epsilon,
+            max_trials,
+            max_epochs,
+            learning_rate_schedule,
+            batch_size,
+            clip_values,
+            verbose,
+        )
+        self.indices_target = indices_target
+        self.selection_strategy = selection_strategy
+        self.patching_strategy = patching_strategy
+        self.retraining_factor = retraining_factor
+        self.model_retrain = model_retrain
+        self.model_retraining_epoch = model_retraining_epoch
+        self.indices_poison = List[int]
+        self.patch = patch
+
+    def poison(
+        self, x_trigger: np.ndarray, y_trigger: np.ndarray, x_train: np.ndarray, y_train: np.ndarray
+    ) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Optimizes a portion of poisoned samples from x_train to make a model classify x_target
+        as y_target by matching the gradients.
+
+        :param x_trigger: A list of samples to use as triggers.
+        :param y_trigger: A list of target classes to classify the triggers into.
+        :param x_train: A list of training data to poison a portion of.
+        :param y_train: A list of labels for x_train.
+        :return: x_train, y_train and indices of poisoned samples.
+                 Here, x_train are the samples selected from target class
+                 in training data.
+        """
+        from art.estimators.classification.pytorch import PyTorchClassifier
+        from art.estimators.classification.tensorflow import TensorFlowV2Classifier
+
+        if isinstance(self.substitute_classifier, TensorFlowV2Classifier):
+            poisoner = self._poison__tensorflow
+            finish_poisoning = self._finish_poison_tensorflow
+            initializer = self._initialize_poison_tensorflow
+        elif isinstance(self.substitute_classifier, PyTorchClassifier):
+            poisoner = self._poison__pytorch
+            finish_poisoning = self._finish_poison_pytorch
+            initializer = self._initialize_poison_pytorch
+        else:
+            raise NotImplementedError("SleeperAgentAttack is currently implemented only for Tensorflow V2 and Pytorch.")
```

Comment:
I think we should move this exception up to the case checking for TensorFlow v2.

Reply:
Agreed

Reply:
Done